### PR TITLE
HeaderDictionaryTypeExtensions not using Dictionary

### LINKED ISSUE #47569: Use FrozenDictionary where appropriate
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

.NET 8 adds `FrozenDictionary`. It's a read-only dictionary with a higher creation cost but better read performance. Good for lookup dictionaries that live a long time.

See https://github.com/dotnet/runtime/issues/67209

### Describe the solution you'd like

Look to see where dictionaries are used today. Consider replacing them with `FrozenDictionary`.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ladeak**
# HeaderDictionaryTypeExtensions not using Dictionary

Considering using FrozenDictionary in `HeaderDictionaryTypeExtensions` - as the Dictionary for parsers does not change. However, measurements show that using `if-else` results in a better overall performance.

## Description

Based on the discussion and performance measurements in #47569, instead of using a `FrozenDictionary` this PR uses `if-else` branching to match known types and known list types when parsing header values.

This way both the *creation* of the dictionary is faster, as we don't have to create one, as well as the lookup.

### CODE REVIEW

**Comment by amcasey** [comment]
This shouldn't actually close https://github.com/dotnet/aspnetcore/issues/47569, should it?  Or does it reflect a comprehensive review of dictionaries that could be converted?

**Comment by ladeak** [comment]
Description updated.

File: src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs [hunk 1]
```diff
@@ -132,45 +132,93 @@ internal static void SetDate(this IHeaderDictionary headers, string name, DateTi
...
         {
             return default(T);
         }
 
-        if (KnownParsers.TryGetValue(typeof(T), out var temp))
+        object? temp = null;
+#pragma warning disable CS8974 // Converting method group to non-delegate type
+        if (typeof(T) == typeof(CacheControlHeaderValue))
+        {
+            temp = ParseCacheControlHeaderValue;
```
Comment (amcasey) [inline src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:182]: Why not `return ParseCacheControlHeaderValue(value.ToString());`?  What's the benefit of wrapping it in `Func` and storing it?

Comment (ladeak) [reply src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:182]: That gives an error: `CS0029	Cannot implicitly convert type 'Microsoft.Net.Http.Headers.CacheControlHeaderValue' to 'T'	Microsoft.AspNetCore.Http.Extensions`

Comment (amcasey) [reply src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:182]: Personally, I'm fine with the casts, especially if it means we can get rid of the pragmas.  Obviously, confirm that it doesn't hurt perf.

Comment (ladeak) [reply src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs]: I don't see a tangible difference in performance, updated as suggested.

File: src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs [hunk 2]
```diff
@@ -132,45 +132,85 @@ internal static void SetDate(this IHeaderDictionary headers, string name, DateTi
...
-
         var value = headers[name];
 
         if (StringValues.IsNullOrEmpty(value))
         {
             return default(T);
         }
 
-        if (KnownParsers.TryGetValue(typeof(T), out var temp))
+        if (typeof(T) == typeof(CacheControlHeaderValue))
```
Comment (amcasey) [inline src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:178]: Personally, I might extract a local for `value.ToString()`.

Comment (amcasey) [reply src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:178]: I might also put a comment in front of the if explaining that it was structured this way for perf, so someone doesn't come along and "clean it up".

**Review by amcasey — APPROVED** [review]
LGTM, but it would be nice to hear from someone on the original thread.  Maybe @BrennanConroy?

File: src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs (hunk 2 above)
Comment (JamesNK) [reply src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:178]: Something like?
```suggestion
        // Comparison of typeof(T) is optimized by JIT.
        if (typeof(T) == typeof(CacheControlHeaderValue))
```

Comment (ladeak) [reply src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:180]: I added the comments and hoisted the `ToString()` - that was very evident not sure how I missed that (...copilot suggestions).

**Review by JamesNK — APPROVED** [review]

**Comment by ladeak** [comment]
Is there anything else I may add before merging?

**Comment by amcasey** [comment]
> Is there anything else I may add before merging?

Nope, LGTM.  Thanks, @ladeak!

File: src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs [hunk 3]
```diff
@@ -132,46 +132,89 @@ internal static void SetDate(this IHeaderDictionary headers, string name, DateTi
...
+        {
+            return (T?)(object?)ParseCacheEntityTagHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(DateTimeOffset?))
+        {
+            return (T?)(object?)ParseCacheDateTimeOffset(value);
+        }
+        else if (typeof(T) == typeof(long?))
+        {
+            return (T?)(object?)ParseCacheInt64(value);
```
Comment (JamesNK) [inline src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:214]: This casts a long to object. Does it box, or is this just ceremony to keep the compiler happy?

Comment (ladeak) [reply src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs:214]: It was to keep compiler happy. Not sure önt eh boxing, I do not recall seeing allocations. (cannot confirm as on the way to an airport)

**Pull request merged**